### PR TITLE
[WIP] 58 add goggle

### LINKED ISSUE #58: Add GOGGLE

### PR DESCRIPTION
**Pull request by robsdavis**
## Description
Added synthetic data model "goggle", this version does not support conditional generation. Closes #58 .
Goggle paper: "GOGGLE: Generative Modelling for Tabular Data by Learning Relational Structure" Authors: Tennison Liu, Zhaozhi Qian, Jeroen Berrevoets, Mihaela van der Schaar

## Affected Dependencies
  - dgl
  - torch_geometric
  - torch_sparse
  - torch_scatter

## How has this been tested?
- Created test_goggle.py with usual plugin tests:
        - test_plugin_sanity
        - test_plugin_name
        - test_plugin_type
        - test_plugin_hyperparams
        - test_plugin_fit
        - test_plugin_generate
        - test_plugin_generate_constraints_goggle
        - test_sample_hyperparams
        - test_eval_performance_goggle
        - test_plugin_encoding
- tests can be run with `pytest -vvvs tests/plugins/generic/test_goggle.py`

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/vanderschaarlab/.github/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/vanderschaarlab/.github/blob/main/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [van der Schaar Lab Styleguide](https://github.com/vanderschaarlab/.github/blob/main/STYLEGUIDE.md)
- [x] I have labelled this PR with the relevant [Type labels](https://github.com/vanderschaarlab/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: src/synthcity/plugins/core/models/RGCNConv.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+        num_bases: Optional[int] = None,
+        num_blocks: Optional[int] = None,
+        aggr: str = "mean",
+        root_weight: bool = True,
+        is_sorted: bool = False,
+        bias: bool = True,
+        **kwargs: Any,
+    ) -> None:
+        kwargs.setdefault("aggr", aggr)
+        super().__init__(node_dim=0, **kwargs)
```
Comment (bcebere) [inline src/synthcity/plugins/core/models/RGCNConv.py:102]: Is this `node_dim` a parameter or needs to be hardcoded to 3?

File: src/synthcity/plugins/core/models/tabular_goggle.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        X_enc = self.encode(X)
+        self.model.fit(
+            X_enc,
+            optimiser_gl=self.optimiser_gl,
+            optimiser_ga=self.optimiser_ga,
+            optimiser=self.optimiser,
+            **kwargs,
+        )
+
+    def enforce_constraints(self, X_synth: torch.Tensor) -> np.ndarray:
```
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_goggle.py:166]: Is this still needed?

File: src/synthcity/plugins/core/models/tabular_goggle.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+# synthcity absolute
+from synthcity.utils.constants import DEVICE
+
+# synthcity relative
+from .goggle import Goggle, GoggleLoss
+from .tabular_encoder import TabularEncoder
+
+
+class TabularGoggle:
```
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_goggle.py:18]: Please add dosctrings for the parameters, some parameters are confusing. For example, `schema` seems to be the feature names.

File: src/synthcity/plugins/generic/plugin_goggle.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+from synthcity.plugins.core.schema import Schema
+from synthcity.utils.constants import DEVICE
+
+
+class GOGGLEPlugin(Plugin):
+    """
+    .. inheritance-diagram:: synthcity.plugins.generic.plugin_GOGGLE.GOGGLEPlugin
+        :parts: 1
+
+    Args:
```
Comment (bcebere) [inline src/synthcity/plugins/generic/plugin_goggle.py:36]: Please document the arguments

File: tests/plugins/generic/test_goggle.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        test_plugin.fit(X)
+        X_syn = test_plugin.generate()
+
+        results.append(evaluator.evaluate(X, X_syn)["syn_id"])
+
+    print(plugin.name(), compress_dataset, decoder_arch, results)
+    assert np.mean(results) > 0.7
+
+
+# @pytest.mark.skipif(sys.platform != "linux", reason="Linux only for faster results")
```
Comment (bcebere) [inline tests/plugins/generic/test_goggle.py:156]: If there are no plans for conditional generation, you should delete these comments.

**Review by bcebere — CHANGES_REQUESTED** [review]

File: src/synthcity/plugins/core/models/tabular_goggle.py (hunk 2 above)
Comment (robsdavis) [reply src/synthcity/plugins/core/models/tabular_goggle.py:166]: Nope, will delete in next push.

File: tests/plugins/generic/test_goggle.py (hunk 5 above)
Comment (robsdavis) [reply tests/plugins/generic/test_goggle.py:156]: I am hoping to add conditional generation as a later date, but will delete in the next push as I haven't seen the research code for it yet.

File: src/synthcity/plugins/core/models/tabular_goggle.py (hunk 3 above)
Comment (robsdavis) [reply src/synthcity/plugins/core/models/tabular_goggle.py:18]: Added docstring. Removed schema argument, as it wasn't required.

File: src/synthcity/plugins/generic/plugin_goggle.py (hunk 4 above)
Comment (robsdavis) [reply src/synthcity/plugins/generic/plugin_goggle.py:37]: Added docstring

File: src/synthcity/plugins/core/models/RGCNConv.py (hunk 1 above)
Comment (robsdavis) [reply src/synthcity/plugins/core/models/RGCNConv.py]: I have exposed it as a parameter now.

**Review by bcebere — APPROVED** [review]
LGTM

File: tests/plugins/generic/test_goggle.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    X = get_airfoil_dataset()
+    test_plugin.fit(GenericDataLoader(X))
+
+
+@pytest.mark.parametrize(
+    "test_plugin", generate_fixtures(plugin_name, plugin, plugin_args)
+)
+@pytest.mark.parametrize("serialize", [True, False])
+def test_plugin_generate(test_plugin: Plugin, serialize: bool) -> None:
+    X = get_airfoil_dataset()
```
[the remark this answers is not in the document]
Comment (robsdavis) [reply tests/plugins/generic/test_goggle.py:62]: I will have a look

File: .github/workflows/test_full.yml [hunk 7]
```diff
@@ -28,6 +28,7 @@ jobs:
       - name: Install dependencies
         run: |
             pip install -r prereq.txt
+            pip install .[goggle]
```
[the remark this answers is not in the document]
Comment (robsdavis) [reply .github/workflows/test_full.yml:31]: Talking with Evgeny, this seems like a better way to handle the dependencies. This issue is that `torch_scatter` and `torch_sparse` need to be installed from source with an --find-links flag or similar. These dependencies are very tricky

File: tests/plugins/generic/test_goggle.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (robsdavis) [reply tests/plugins/generic/test_goggle.py]: Swapped for diabetes

File: src/synthcity/plugins/core/plugin.py [hunk 8]
```diff
@@ -548,8 +548,22 @@ def __init__(self, plugins: list, expected_type: Type, categories: list) -> None
         self._available_plugins = {}
         for plugin in plugins:
             stem = Path(plugin).stem.split("plugin_")[-1]
-            self._available_plugins[stem] = plugin
-
+            if stem == "goggle":
```
Comment (bcebere) [inline src/synthcity/plugins/core/plugin.py:551]: Please don't include specific plugin conditions in `plugin.py`, this should stay generic.

**Review by bcebere — CHANGES_REQUESTED** [review]

**Review by bcebere — APPROVED** [review]
